IP address & port w/o scheme in endpoint ⇒ URISyntaxException/IllegalArgumentException from EndpointToRegion.guessRegionOrRegionNameForEndpoint

<!--- Provide a general summary of the issue in the Title above -->

## Describe the bug
<!--- A clear and concise description of what the bug is -->

If you pass a custom endpoint in the format `127.0.0.1:1234` to the SDK, the client builder will fail.

## Expected Behavior

Should behave similarly to if a hostname such as `localhost` is used, or if an IP address is used without an explicit port, or if a full URL with scheme is passed.

## Current Behavior

You get an error such as

```
java.lang.IllegalArgumentException: Illegal character in scheme name at index 0: 127.0.0.1:38835
	at java.base/java.net.URI.create(URI.java:883)
	at com.amazonaws.regions.EndpointToRegion.guessRegionOrRegionNameForEndpoint(EndpointToRegion.java:71)
	at com.amazonaws.regions.EndpointToRegion.guessRegionNameForEndpoint(EndpointToRegion.java:41)
	at com.amazonaws.AmazonWebServiceClient.setEndpoint(AmazonWebServiceClient.java:323)
	at com.amazonaws.services.s3.AmazonS3Client.setEndpoint(AmazonS3Client.java:725)
	at com.amazonaws.client.builder.AwsClientBuilder.setRegion(AwsClientBuilder.java:453)
	at com.amazonaws.client.builder.AwsClientBuilder.configureMutableProperties(AwsClientBuilder.java:424)
	at com.amazonaws.client.builder.AwsSyncClientBuilder.build(AwsSyncClientBuilder.java:46)
	at …
Caused by: java.net.URISyntaxException: Illegal character in scheme name at index 0: 127.0.0.1:38835
	at java.base/java.net.URI$Parser.fail(URI.java:2913)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3084)
	at java.base/java.net.URI$Parser.checkChar(URI.java:3094)
	at java.base/java.net.URI$Parser.parse(URI.java:3109)
	at java.base/java.net.URI.<init>(URI.java:600)
	at java.base/java.net.URI.create(URI.java:881)
	... 23 more
```

<!--- Tell us what happens instead of the expected behavior -->

<!--- Include full errors, uncaught exceptions, stack traces, and relevant logs -->
<!--- To turn on SDK logging, follow instructions here: https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/java-dg-logging.html-->
<!--- If service responses are relevant, please include wirelogs -->

## Steps to Reproduce
<!--- Provide a self-contained, concise snippet of code that can be used to reproduce the issue -->
<!--- For more complex issues provide a repo with the smallest sample that reproduces the bug -->
<!--- Avoid including business logic or unrelated code, it makes diagnosis more difficult -->

Reproduced in https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/112.

## Possible Solution
<!--- Not required, but suggest a fix/reason for the bug -->

PR forthcoming.

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

Broke integration testing of the Jenkins plugin to store artifacts in S3 against a MinIO mock server running in a local Docker container.
The test passed against 1.11.700 and failed against 1.11.723, with the introduction of `EndpointToRegion`.

## Your Environment
<!--- Include as many relevant details about the environment where the bug was discovered -->
* AWS Java SDK version used: 1.11.700 vs. 1.11.723
* JDK version used: 8 or 11
* Operating System and version: various Linux

--- a/aws-java-sdk-core/src/main/java/com/amazonaws/regions/EndpointToRegion.java
+++ b/aws-java-sdk-core/src/main/java/com/amazonaws/regions/EndpointToRegion.java
@@ -17,6 +17,7 @@ package com.amazonaws.regions;
 import com.amazonaws.annotation.SdkProtectedApi;
 import com.amazonaws.util.AwsHostNameUtils;
 import java.net.URI;
+import java.net.URISyntaxException;
 
 /**
  * Utilities to attempt to convert from a hostname/endpoint to an AWS region.
@@ -68,7 +69,12 @@ public class EndpointToRegion {
             return new RegionOrRegionName();
         }
 
-        String host = URI.create(endpoint).getHost();
+        String host = null;
+        try {
+            host = new URI(endpoint).getHost();
+        } catch (URISyntaxException x) {
+            // proceed
+        }
         if (host == null) {
             host = URI.create("http://" + endpoint).getHost();
         }

--- a/aws-java-sdk-core/src/test/java/com/amazonaws/regions/EndpointToRegionTest.java
+++ b/aws-java-sdk-core/src/test/java/com/amazonaws/regions/EndpointToRegionTest.java
@@ -111,6 +111,18 @@ public class EndpointToRegionTest {
         verifyRegionAndPartitionForHostname("us-gov-west-1", "aws-us-gov", "iam.us-gov-west-1.banana.com", "iam");
     }
 
+    @Test
+    public void guessRegionForHostname_ipAddress() {
+        assertNull(guessRegionNameForEndpoint("http://localhost"));
+        assertNull(guessRegionNameForEndpoint("http://localhost:1234"));
+        assertNull(guessRegionNameForEndpoint("http://127.0.0.1"));
+        assertNull(guessRegionNameForEndpoint("http://127.0.0.1:1234"));
+        assertNull(guessRegionNameForEndpoint("localhost"));
+        assertNull(guessRegionNameForEndpoint("localhost:1234"));
+        assertNull(guessRegionNameForEndpoint("127.0.0.1"));
+        assertNull(guessRegionNameForEndpoint("127.0.0.1:1234"));
+    }
+
     /**
      * We migrated from AwsHostNameUtils.parseRegion to EndpointToRegion.guessRegionNameForHostname, because EndpointToRegion
      * can consider the contents of endpoints.json.

